lib/decision/tls_identity.go

```diff
@@ -0,0 +1,273 @@
+		PrivateKeyPolicy:        keys.PrivateKeyPolicy(id.PrivateKeyPolicy),
+		ConnectionDiagnosticID:  id.ConnectionDiagnosticId,
+		DeviceExtensions:        deviceExtensionsFromProto(id.DeviceExtensions),
+		UserType:                types.UserType(id.UserType),
+	}
+}
+
+// TLSIdentityFromTLSCA transforms a [tlsca.Identity] into its equivalent
+// [decisionpb.TLSIdentity].
+func TLSIdentityFromTLSCA(id *tlsca.Identity) *decisionpb.TLSIdentity {
+	if id == nil {
+		return nil
+	}
+
+	return &decisionpb.TLSIdentity{
+		Username:                id.Username,
+		Impersonator:            id.Impersonator,
+		Groups:                  id.Groups,
+		SystemRoles:             id.SystemRoles,
+		Usage:                   id.Usage,
+		Principals:              id.Principals,
+		KubernetesGroups:        id.KubernetesGroups,
+		KubernetesUsers:         id.KubernetesUsers,
+		Expires:                 timestampFromGoTime(id.Expires),
+		RouteToCluster:          id.RouteToCluster,
+		KubernetesCluster:       id.KubernetesCluster,
+		Traits:                  traitFromWrappers(id.Traits),
+		RouteToApp:              routeToAppToProto(&id.RouteToApp),
+		TeleportCluster:         id.TeleportCluster,
+		RouteToDatabase:         routeToDatabaseToProto(&id.RouteToDatabase),
+		DatabaseNames:           id.DatabaseNames,
+		DatabaseUsers:           id.DatabaseUsers,
+		MfaVerified:             id.MFAVerified,
+		PreviousIdentityExpires: timestampFromGoTime(id.PreviousIdentityExpires),
+		LoginIp:                 id.LoginIP,
+		PinnedIp:                id.PinnedIP,
+		AwsRoleArns:             id.AWSRoleARNs,
+		AzureIdentities:         id.AzureIdentities,
+		GcpServiceAccounts:      id.GCPServiceAccounts,
+		ActiveRequests:          id.ActiveRequests,
+		DisallowReissue:         id.DisallowReissue,
+		Renewable:               id.Renewable,
+		Generation:              id.Generation,
+		BotName:                 id.BotName,
+		BotInstanceId:           id.BotInstanceID,
+		AllowedResourceIds:      resourceIDsFromTypes(id.AllowedResourceIDs),
+		PrivateKeyPolicy:        string(id.PrivateKeyPolicy),
+		ConnectionDiagnosticId:  id.ConnectionDiagnosticID,
+		DeviceExtensions:        deviceExtensionsToProto(&id.DeviceExtensions),
+		UserType:                string(id.UserType),
+	}
+}
+
+func timestampToGoTime(t *timestamppb.Timestamp) time.Time {
+	if t == nil {
+		return time.Time{}
+	}
+	return t.AsTime()
+}
+
+func timestampFromGoTime(t time.Time) *timestamppb.Timestamp {
+	if t.IsZero() {
+		return nil
+	}
+	return timestamppb.New(t)
+}
+
+func traitToWrappers(traits []*traitpb.Trait) wrappers.Traits {
+	apiTraits := apitraitconvert.FromProto(traits)
+	return wrappers.Traits(apiTraits)
+}
+
+func traitFromWrappers(traits wrappers.Traits) []*traitpb.Trait {
+	if len(traits) == 0 {
+		return nil
+	}
+	apiTraits := apitrait.Traits(traits)
+	return apitraitconvert.ToProto(apiTraits)
+}
+
+func routeToAppFromProto(routeToApp *decisionpb.RouteToApp) tlsca.RouteToApp {
```

Comment:
This returns tlsca.RouterToApp by value as that is how tlsca.Identity wants it. The inverse conversion (routeToAppToProto) will both take and return a pointer.

Same for other structs.